src/Stack/Ls.hs

```diff
@@ -0,0 +1,219 @@
+    = Local
+    | Remote
+    deriving (Show, Eq, Ord)
+
+data SnapshotType
+    = Lts
+    | Nightly
+    deriving (Show, Eq, Ord)
+
+data LsCmds =
+    LsSnapshot SnapshotOpts
+    deriving (Eq, Show, Ord)
+
+data SnapshotOpts = SnapshotOpts
+    { soptViewType :: LsView
+    , soptLtsSnapView :: Bool
+    , soptNightlySnapView :: Bool
+    } deriving (Eq, Show, Ord)
+
+data LsCmdOpts = LsCmdOpts
+    { lsView :: LsCmds
+    } deriving (Eq, Show, Ord)
+
+lsParser :: OA.Parser LsCmdOpts
+lsParser = LsCmdOpts <$> OA.hsubparser lsSnapCmd
+
+lsCmdOptsParser :: OA.Parser LsCmds
+lsCmdOptsParser = fmap LsSnapshot lsViewSnapCmd
+
+lsViewSnapCmd :: OA.Parser SnapshotOpts
+lsViewSnapCmd =
+    SnapshotOpts <$> OA.hsubparser (lsViewRemoteCmd <> lsViewLocalCmd) <*>
+    OA.switch
+        (OA.long "lts" <> OA.short 'l' <> OA.help "Only show lts snapshots") <*>
+    OA.switch
+        (OA.long "nightly" <> OA.short 'n' <>
+         OA.help "Only show nightly snapshots")
+
+lsSnapCmd :: OA.Mod OA.CommandFields LsCmds
+lsSnapCmd =
+    OA.command
+        "snapshots"
+        (OA.info lsCmdOptsParser (OA.progDesc "View local snapshot"))
+
+data Snapshot = Snapshot
+    { snapId :: Text
+    , snapTitle :: Text
+    , snapTime :: Text
+    } deriving (Show, Eq, Ord)
+
+data SnapshotData = SnapshotData
+    { _snapTotalCounts :: Integer
+    , snaps :: [[Snapshot]]
+    } deriving (Show, Eq, Ord)
+
+instance FromJSON Snapshot where
+    parseJSON o@(Array _) = parseSnapshot o
+    parseJSON _ = mempty
+
+instance FromJSON SnapshotData where
+    parseJSON (Object s) =
+        SnapshotData <$> s .: "totalCount" <*> s .: "snapshots"
+    parseJSON _ = mempty
+
+toSnapshot :: [Value] -> Snapshot
+toSnapshot [String sid, String stitle, String stime] =
+    Snapshot
+    { snapId = sid
+    , snapTitle = stitle
+    , snapTime = stime
+    }
+toSnapshot val = throw $ ParseFailure val
+
+newtype LsException =
+    ParseFailure [Value]
+    deriving (Show, Typeable)
+
+instance Exception LsException
+
+parseSnapshot :: Value -> A.Parser Snapshot
+parseSnapshot = A.withArray "array of snapshot" (return . toSnapshot . V.toList)
+
+displayTime :: Snapshot -> [ByteString]
+displayTime Snapshot {..} = [pack $ T.unpack snapTime]
+
+displaySnap :: Snapshot -> [ByteString]
+displaySnap Snapshot {..} =
+    [ "Resolver name: " <> pack (T.unpack snapId)
+    , "\n" <> pack (T.unpack snapTitle) <> "\n\n"
+    ]
+
+displaySingleSnap :: [Snapshot] -> ByteString
+displaySingleSnap snapshots =
+    case snapshots of
+        [] -> mempty
+        (x:xs) ->
+            let snaps =
+                    displayTime x <> ["\n\n"] <> displaySnap x <>
+                    L.concatMap displaySnap xs
+            in BC.concat snaps
+
+displaySnapshotData :: SnapshotData -> IO ()
+displaySnapshotData sdata =
+    case L.reverse $ snaps sdata of
+        [] -> return ()
+        xs ->
+            let snaps = BC.concat $ L.map displaySingleSnap xs
+            in pageByteString snaps
```

Comment:
There's an issue with always using the pager, which is that someone might want to consume the output programmatically via stdout.  Perhaps do `isStdoutTerminal <- view terminalL` and if true, then run the pager?  If false, then assume it's running in a script, and instead send to stdout.

May also be worth considering representing this stuff as Text not ByteString.  In a few places above `pack` is used to go from String to ByteString.  I think this is fine, because the Strings will always be ASCII, but it would break on unicode.

Reply:
Thanks, will do this.